crates/ruff_linter/src/rules/pycodestyle/rules/logical_lines/blank_lines.rs

```diff
@@ -0,0 +1,620 @@
+#[violation]
+pub struct BlankLinesBeforeNestedDefinition(pub u32);
+
+impl AlwaysFixableViolation for BlankLinesBeforeNestedDefinition {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BlankLinesBeforeNestedDefinition(blank_lines) = self;
+        format!("Expected 1 blank line before a nested definition, found {blank_lines}")
+    }
+
+    fn fix_title(&self) -> String {
+        "Add missing blank line".to_string()
+    }
+}
+
+/// Returns `true` if line is a docstring only line.
+fn is_docstring(line: Option<&LogicalLine>) -> bool {
+    line.is_some_and(|line| {
+        !line.tokens_trimmed().is_empty()
+            && line
+                .tokens_trimmed()
+                .iter()
+                .all(|token| matches!(token.kind(), TokenKind::String))
+    })
+}
+
+/// Returns `true` if the token is Async, Class or Def
+fn is_top_level_token(token: Option<TokenKind>) -> bool {
+    token.is_some_and(|token| matches!(token, TokenKind::Class | TokenKind::Def | TokenKind::Async))
+}
+
+/// Returns `true` if the token is At, Async, Class or Def
+fn is_top_level_token_or_decorator(token: TokenKind) -> bool {
+    matches!(
+        token,
+        TokenKind::Class | TokenKind::Def | TokenKind::Async | TokenKind::At
+    )
+}
+
+/// E301, E302, E303, E304, E305, E306
+#[allow(clippy::too_many_arguments)]
+pub(crate) fn blank_lines(
+    line: &LogicalLine,
+    tracked_vars: &mut BlankLinesTrackingVars,
+    prev_indent_level: Option<usize>,
+    indent_level: usize,
+    indent_size: usize,
+    locator: &Locator,
+    stylist: &Stylist,
+    context: &mut LogicalLinesContext,
+) {
+    let line_is_comment_only = line.is_comment_only();
+
+    if indent_level < tracked_vars.class_indent_level && tracked_vars.is_in_class {
+        tracked_vars.is_in_class = false;
+        if line_is_comment_only {
+            tracked_vars.follows_comment_after_class = true;
+        }
+    }
+
+    if indent_level < tracked_vars.fn_indent_level && tracked_vars.is_in_fn {
+        tracked_vars.is_in_fn = false;
+        if line_is_comment_only {
+            tracked_vars.follows_comment_after_fn = true;
+        }
+    }
+
+    // A comment can be de-indented while still being in a class/function, in that case
+    // we need to revert the variables.
+    if tracked_vars.follows_comment_after_fn && !line_is_comment_only {
+        if indent_level == tracked_vars.fn_indent_level {
+            tracked_vars.is_in_fn = true;
+        }
+        tracked_vars.follows_comment_after_fn = false;
+    }
+
+    if tracked_vars.follows_comment_after_class && !line_is_comment_only {
+        if indent_level == tracked_vars.class_indent_level {
+            tracked_vars.is_in_class = true;
+        }
+        tracked_vars.follows_comment_after_class = false;
+    }
+
+    for token in line.tokens() {
+        if matches!(token.kind, TokenKind::Indent | TokenKind::Dedent) {
+            continue;
+        }
+
```

Comment:
Nit: I think the code is easier to read when you remove the loop to reduce the nestin. 

```rust
let first_token = line.tokens().find(|token| !matches!(token.kind, TokenKind::Indent | TokenKind::Dedent));

// The body of the loop
```

It makes it clear which tokens are skipped and that the body itself is only executed once (it also removes the need for the `break` statements in the `match`` below)